Add visual indication of lost logs to UART and RTT backends

nordic-krch opened the pull request:
Added implementation of new interface added in #11769.
It touches #11763 as it fixes it for UART, RTT backends. native_posix and shell is still not implementing it.

When messages are dropped `~` is printed. Number of printed `~` indicates number of dropped messages, giving rough idea how much buffer should be increased.

nordic-krch committed:
logging: Add handling of dropped messages to UART backend

Extend UART backend to handle dropped messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add handling of dropped messages to RTT backend

Extend RTT backend to handle dropped messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

jhedberg commented on subsys/logging/log_backend_uart.c:
```diff
@@ -62,10 +62,24 @@ static void panic(struct log_backend const *const backend)
+		uart_poll_out(dev, '~');
+	}
+
+	char_out((u8_t *)newline_str, sizeof(newline_str), dev);
```
sizeof(newline_str) evaluates to 3, is it intentional that you're also writing the terminating null byte?

nordic-krch replied:
right, sending null wasn't intentional. Fixed to send sizeof - 1.

pfalcon commented:
Well, that's definitely a step in the right direction, but why not print clearly: `~Dropped: NN\n` ?

pfalcon commented:
In particular, why print a thousand of `~`, if 1000 messages were dropped?

pizi-nordic reviewed (commented):
I agree with @pfalcon, that printing "--- %u messages lost ---" will be better. However I understand, that we should avoid printing too much, as we should use the bandwidth for log messages.

nordic-krch commented:
ok, i agree that message might be better but it i would add red coloring to it as it must be spotted by the user.

nordic-krch committed:
logging: Handle dropped messages in log_output

Extended log_output interface to handle dropped
messages. Log_output is printing a message containing
number of dropped messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Add handling of dropped messages to native POSIX backend

Extended native POSIX backend to handle dropped messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
please re-review as I've reworked the thing:
- added generic handling of dropped messages in log_output
- backends are using log_output function so output will be uniform across backends
- adding dropped message support to the backend is trivial
- message `--- nn messages dropped ---` is red if coloring is enabled

pfalcon commented on subsys/logging/log_output.c:
```diff
@@ -522,6 +528,24 @@ void log_output_msg_process(const struct log_output *log_output,
+	log_output_func_t outf = log_output->func;
+	struct device *dev = (struct device *)log_output->control_block->ctx;
+
+	cnt = min(cnt, 99);
```
But why? What's wrong with printing 5-digit number?

nordic-krch replied:
increased to 4 digits. It's unlikely to see more than 9999 lost messages and even then exact number is not important.

pfalcon commented on subsys/logging/log_output.c:
```diff
@@ -114,18 +115,23 @@ static int print_formatted(const struct log_output *log_output,
 }
 
-void log_output_flush(const struct log_output *log_output)
+static void buffer_flush(log_output_func_t outf, u8_t *buf, size_t len,
```
I find it strange that function which performs "write" operation is called "flush". "Flush" operation generally doesn't take params (except for maybe timeout), and flushes whatever already accumulated.

nordic-krch replied:
yes, i did rename only function prefix after extracting it from `log_output_flush`. changed to `buffer_write`.

pfalcon reviewed (approved):
Thanks for the updates!

pfalcon commented:
@jhedberg, @pizi-nordic, @jarz-nordic, @carlescufi: Can you please (re)review this, so we paced forward with logging fixes?
